module/zfs/dsl_scan.c

```diff
@@ -2184,52 +2184,75 @@ dsl_scan_ds_clone_swapped(dsl_dataset_t *ds1, dsl_dataset_t *ds2, dmu_tx_t *tx)
 {
 	dsl_pool_t *dp = ds1->ds_dir->dd_pool;
 	dsl_scan_t *scn = dp->dp_scan;
-	uint64_t mintxg;
+	uint64_t mintxg1, mintxg2;
+	boolean_t ds1_queued, ds2_queued;
 
 	if (!dsl_scan_is_running(scn))
 		return;
 
 	ds_clone_swapped_bookmark(ds1, ds2, &scn->scn_phys.scn_bookmark);
 	ds_clone_swapped_bookmark(ds1, ds2, &scn->scn_phys_cached.scn_bookmark);
 
-	if (scan_ds_queue_contains(scn, ds1->ds_object, &mintxg)) {
-		scan_ds_queue_remove(scn, ds1->ds_object);
-		scan_ds_queue_insert(scn, ds2->ds_object, mintxg);
+	/*
+	 * Handle the in-memory scan queue.
+	 */
+	ds1_queued = scan_ds_queue_contains(scn, ds1->ds_object, &mintxg1);
+	ds2_queued = scan_ds_queue_contains(scn, ds2->ds_object, &mintxg2);
+
+	/* Sanity checking. */
+	if (ds1_queued) {
+		ASSERT3U(mintxg1, ==, dsl_dataset_phys(ds1)->ds_prev_snap_txg);
+		ASSERT3U(mintxg1, ==, dsl_dataset_phys(ds2)->ds_prev_snap_txg);
 	}
-	if (scan_ds_queue_contains(scn, ds2->ds_object, &mintxg)) {
+	if (ds2_queued) {
+		ASSERT3U(mintxg2, ==, dsl_dataset_phys(ds1)->ds_prev_snap_txg);
+		ASSERT3U(mintxg2, ==, dsl_dataset_phys(ds2)->ds_prev_snap_txg);
+	}
+
+	if (ds1_queued && ds2_queued) {
+		/* Nothing to do if both are queued (or both are not queued). */
```

Comment:
How about changing this comment to:
`If both are queued, we don't need to do anything.  The swapping code below would not handle this case correctly, since we can't insert ds2 if it is already there.`

Reply:
Thank you for the suggestion! I will take it.
At the same time it's not the only problem with the hypothetical alternative mentioned in the comment. It would attempt to replace ds1 with ds2 but not in the opposite direction because of `else if`. And without the `else if` there would be another problem: replacing ds1 with ds2 and then ds2 back with ds1.
So, in my opinion, even if the queue silently ignored duplicate inserts, it would be very hard (if possible) to handle the case where both ds1 and ds2 are queued without explicitly handling that case in one way or another.
E.g., the proposed code shows one way; the existing code for the on-disk queue shows another: having the enqueue operation return an error code, checking for an error, and if the error is EEXIST, then backing out the original removal. Which is more work than just doing nothing in the first place.